File: app/src/main/java/com/vegasoft/mypasswords/presentation/fragments/RecordListFragment.java
package com.vegasoft.mypasswords.presentation.fragments;

import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.vegasoft.mypasswords.R;
import com.vegasoft.mypasswords.data.db.RecordsDBHelper;
import com.vegasoft.mypasswords.data.entity.Record;
import com.vegasoft.mypasswords.data.loader.LoaderResult;
import com.vegasoft.mypasswords.data.loader.RecordsLoader;
import com.vegasoft.mypasswords.presentation.OnListFragmentInteractionListener;
import com.vegasoft.mypasswords.presentation.adapter.RecordAdapter;

import java.util.ArrayList;

public class RecordListFragment extends Fragment {

    private static final int RECORD_LOADER_ID = 0;

    private OnListFragmentInteractionListener mListener;
    private RecordAdapter adapter;

    private LoaderManager.LoaderCallbacks<LoaderResult<ArrayList<Record>>> generalLoader = new LoaderManager.LoaderCallbacks<LoaderResult<ArrayList<Record>>>() {

        @Override
        public Loader<LoaderResult<ArrayList<Record>>> onCreateLoader(int arg0,
                                                                      Bundle arg1) {
            return new RecordsLoader(getActivity());
        }

        @Override
        public void onLoadFinished(
                Loader<LoaderResult<ArrayList<Record>>> loader,
                LoaderResult<ArrayList<Record>> result) {
            if (getView() == null) {
                return;
            }
            getView().findViewById(R.id.progressBar).setVisibility(View.GONE);
            if (result.error == null) {
                if (result.data != null) {
                    adapter.setItems(result.data);
                    if (result.data.size() < 1) {
                        getView().findViewById(R.id.empty_message).setVisibility(View.VISIBLE);
                    }
                } else {
                    getView().findViewById(R.id.empty_message).setVisibility(View.VISIBLE);
                }
            }
        }

        @Override
        public void onLoaderReset(Loader<LoaderResult<ArrayList<Record>>> loader) {
        }

    };

    ItemTouchHelper.SimpleCallback simpleItemTouchCallback = new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT) {
        @Override
        public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
            return false;
        }

        @Override
        public float getSwipeThreshold(RecyclerView.ViewHolder viewHolder) {
            return super.getSwipeThreshold(viewHolder);
        }

        @Override
        public void onSwiped(final RecyclerView.ViewHolder viewHolder, int swipeDir) {
            if (getView() == null) {
                return;
            }
            final int adapterPosition = viewHolder.getAdapterPosition();
            final Record record = adapter.getRecord(adapterPosition);
            new RecordsDBHelper(getActivity()).deleteRecord(record.getId());
            adapter.removeItem(adapterPosition);
            final Snackbar snackbar = Snackbar
                    .make(getView().findViewById(R.id.myCoordinatorLayout), "Do you want return record?", Snackbar.LENGTH_SHORT)
                    .setAction("YES", new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            new RecordsDBHelper(getActivity()).addRecord(record);
                            adapter.addItem(adapterPosition, record);
                            getView().findViewById(R.id.empty_message).setVisibility(View.GONE);
                        }
                    });

            snackbar.setActionTextColor(Color.RED);
            View sbView = snackbar.getView();
            TextView textView = sbView.findViewById(android.support.design.R.id.snackbar_text);
            textView.setTextColor(Color.WHITE);
            snackbar.show();
            if (adapter.getItemCount() < 1) {
                getView().findViewById(R.id.empty_message).setVisibility(View.VISIBLE);
            }
        }
    };

    public RecordListFragment() {
    }

    public static RecordListFragment newInstance() {
        return new RecordListFragment();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_pass_list, container, false);

        // Set the adapter
        RecyclerView recyclerView = view.findViewById(R.id.list);
        recyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));
        adapter = new RecordAdapter(mListener);
        recyclerView.setAdapter(adapter);
        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(simpleItemTouchCallback);
        itemTouchHelper.attachToRecyclerView(recyclerView);

        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getLoaderManager().restartLoader(RECORD_LOADER_ID, null,
                generalLoader).forceLoad();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnListFragmentInteractionListener) {
            mListener = (OnListFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnListFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

}

File: app/src/main/java/com/vegasoft/mypasswords/presentation/fragments_new/records/list/RecordListFragment.kt
package com.vegasoft.mypasswords.presentation.fragments_new.records.list

import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.vegasoft.mypasswords.R
import com.vegasoft.mypasswords.presentation.OnListFragmentInteractionListener
import com.vegasoft.mypasswords.presentation.adapter.RecordAdapter
import kotlinx.android.synthetic.main.fragment_record_list.empty_message as emptyMessage
import kotlinx.android.synthetic.main.fragment_record_list.myCoordinatorLayout

class RecordListFragment : Fragment() {

    private var mListener: OnListFragmentInteractionListener? = null
    private lateinit var adapter: RecordAdapter


    var simpleItemTouchCallback: ItemTouchHelper.SimpleCallback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            return false
        }

        override fun getSwipeThreshold(viewHolder: RecyclerView.ViewHolder): Float {
            return super.getSwipeThreshold(viewHolder)
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, swipeDir: Int) {

            val adapterPosition = viewHolder.adapterPosition
            val record = adapter.getRecord(adapterPosition) ?: return
            viewModel.removeRecord(requireContext(), record.id)
            adapter.removeItem(adapterPosition)
            view?.let {
                val snackbar = Snackbar
                        .make(myCoordinatorLayout, "Do you want to restore record?", Snackbar.LENGTH_SHORT)
                        .setAction("YES") {
                            viewModel.addRecord(requireContext(), record)
                            adapter.addItem(adapterPosition, record)
                            emptyMessage.visibility = View.GONE
                        }
                snackbar.setActionTextColor(Color.RED)
                val sbView = snackbar.view
                val textView = sbView.findViewById<TextView>(R.id.snackbar_text)
                textView.setTextColor(Color.WHITE)
                snackbar.show()
                if (adapter.isEmpty) {
                    emptyMessage.visibility = View.VISIBLE
                }
            }
        }
    }

    companion object {
        fun newInstance() = RecordListFragment()
    }

    private lateinit var viewModel: RecordListViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_record_list, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        viewModel = ViewModelProviders.of(this).get(RecordListViewModel::class.java)
        viewModel._records.observe(viewLifecycleOwner, Observer {
            adapter.setItems(it)
        })
        viewModel.error.observe(viewLifecycleOwner, Observer {
            Toast.makeText(
                    requireContext(),
                    it.message,
                    Toast.LENGTH_SHORT
            ).show()
        })
    }

    override fun onStart() {
        super.onStart()
        viewModel.loadRecords(requireContext())
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Set the adapter
        val recyclerView: RecyclerView = view.findViewById(R.id.list)
        recyclerView.layoutManager = LinearLayoutManager(view.context)
        adapter = RecordAdapter(mListener)
        recyclerView.adapter = adapter
        val itemTouchHelper = ItemTouchHelper(simpleItemTouchCallback)
        itemTouchHelper.attachToRecyclerView(recyclerView)
    }


    override fun onAttach(context: Context) {
        super.onAttach(context)
        mListener = if (context is OnListFragmentInteractionListener) {
            context
        } else {
            throw RuntimeException(context.toString()
                    + " must implement OnListFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }
}
